jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.cpp

```diff
@@ -1667,6 +1667,127 @@ ecma_builtin_string_prototype_object_slice (ecma_value_t this_arg, /**< this arg
+ *                            first character in the input_string that matched
+ *
+ *         Returned value must be freed with ecma_free_completion_value.
+ */
+static ecma_completion_value_t
+ecma_builtin_helper_split_match (ecma_value_t input_string, /**< first argument */
+                                 ecma_length_t start_idx, /**< second argument */
+                                 ecma_value_t separator) /**< third argument */
+{
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+
+  /* 1. */
+  if (ecma_is_value_object (separator)
+      && ecma_object_get_class_name (ecma_get_object_from_value (separator)) == LIT_MAGIC_STRING_REGEXP_UL)
+  {
+    ecma_value_t regexp_value = ecma_copy_value (separator, false);
+
+    ECMA_TRY_CATCH (to_string_val,
+                    ecma_op_to_string (input_string),
+                    ret_value);
+
+    ecma_string_t *input_str_p = ecma_get_string_from_value (to_string_val);
+    ecma_string_t *substr_str_p = ecma_string_substr (input_str_p, start_idx, ecma_string_get_length (input_str_p));
+
+    ret_value = ecma_regexp_exec_helper (regexp_value, ecma_make_string_value (substr_str_p), true);
+
+    if (!ecma_is_value_null (ecma_get_completion_value_value (ret_value)))
+    {
+      ecma_object_t *obj_p = ecma_get_object_from_completion_value (ret_value);
+      ecma_string_t *magic_str_p = ecma_get_magic_string (LIT_MAGIC_STRING_INDEX);
+      ecma_property_t *index_prop_p = ecma_find_named_property (obj_p, magic_str_p);
+
+      JERRY_ASSERT (index_prop_p != NULL);
+
+      ecma_number_t *index_num_p = ecma_get_number_from_value (index_prop_p->u.named_data_property.value);
+      *index_num_p += start_idx;
+
+      ecma_deref_ecma_string (magic_str_p);
+    }
+
+    ecma_deref_ecma_string (substr_str_p);
+    ECMA_FINALIZE (to_string_val);
+  }
+  else
+  {
+    /* 2. */
+    JERRY_ASSERT (ecma_is_value_string (input_string) && ecma_is_value_string (separator));
+
+    ecma_string_t *string_str_p = ecma_get_string_from_value (input_string);
+    ecma_string_t *separator_str_p = ecma_get_string_from_value (separator);
+
+    /* 3. */
+    ecma_length_t string_length = ecma_string_get_length (string_str_p);
+    ecma_length_t separator_length = ecma_string_get_length (separator_str_p);
+
+    /* 4. */
+    if (start_idx + separator_length > string_length)
+    {
+      ret_value = ecma_make_simple_completion_value (ECMA_SIMPLE_VALUE_NULL);
+    }
+    else
+    {
+      bool is_different = false;
+
+      /* 5. */
+      for (ecma_length_t i = 0; i < separator_length && !is_different; i++)
+      {
+        ecma_char_t char_from_string = ecma_string_get_char_at_pos (string_str_p, start_idx + i);
+        ecma_char_t char_from_separator = ecma_string_get_char_at_pos (separator_str_p, i);
+
+        if (char_from_string != char_from_separator)
+        {
+          ret_value = ecma_make_simple_completion_value (ECMA_SIMPLE_VALUE_NULL);
```

Comment:
Could you, please, the assignment outside of the loop, and use `is_different` instead of `ecma_is_completion_value_empty`?
Its better to use `ecma_is_completion_value_empty` only for checking exceptions, not for general conditionals.